localization/yabloc/yabloc_particle_filter/src/gnss_corrector/gnss_corrector_core.cpp

```diff
@@ -0,0 +1,221 @@
+  // const Eigen::Vector3f doppler = extract_enu_vel(*ublox_msg);
+  // const float theta = std::atan2(doppler.y(), doppler.x());
+  // add_weight_by_orientation(weighted_particles, doppler);
+
+  // Compute travel distance from last update position
+  // If the distance is too short, skip weighting
+  {
+    Eigen::Vector3f meaned_position = common::pose_to_affine(meaned_pose).translation();
+    if ((meaned_position - last_mean_position_).squaredNorm() > 1) {
+      this->set_weighted_particle_array(weighted_particles);
+      last_mean_position_ = meaned_position;
+    } else {
+      RCLCPP_WARN_STREAM_THROTTLE(
+        get_logger(), *get_clock(), 2000,
+        "Skip particle weighting due to almost the same position");
+    }
+  }
+}
+
+void GnssParticleCorrector::publish_marker(const Eigen::Vector3f & position, bool is_rtk_fixed)
+{
+  using namespace std::literals::chrono_literals;
+  using Point = geometry_msgs::msg::Point;
+
+  auto drawCircle = [](std::vector<Point> & points, float radius) -> void {
+    const int N = 10;
+    for (int theta = 0; theta < 2 * N + 1; theta++) {
+      geometry_msgs::msg::Point pt;
+      pt.x = radius * std::cos(theta * M_PI / N);
+      pt.y = radius * std::sin(theta * M_PI / N);
+      points.push_back(pt);
+    }
+  };
+
+  MarkerArray array_msg;
+  for (int i = 0; i < 5; i++) {
+    Marker marker;
+    marker.header.stamp = get_clock()->now();
+    marker.header.frame_id = "/map";
+    marker.id = i;
+    marker.type = Marker::LINE_STRIP;
+    marker.lifetime = rclcpp::Duration(500ms);
+    marker.pose.position.x = position.x();
+    marker.pose.position.y = position.y();
+    marker.pose.position.z = latest_height_.data;
+
+    float prob = i / 4.0f;
+    marker.color = common::color_scale::rainbow(prob);
+    marker.color.a = 0.5f;
+    marker.scale.x = 0.1;
+    drawCircle(marker.points, weight_manager_.inverse_normal_pdf(prob, is_rtk_fixed));
+    array_msg.markers.push_back(marker);
+  }
+  marker_pub_->publish(array_msg);
+}
+
+GnssParticleCorrector::ParticleArray GnssParticleCorrector::weight_particles(
+  const ParticleArray & predicted_particles, const Eigen::Vector3f & pose, bool is_rtk_fixed)
+{
+  ParticleArray weighted_particles{predicted_particles};
+
+  for (auto & particle : weighted_particles.particles) {
+    float distance = static_cast<float>(
+      std::hypot(particle.pose.position.x - pose.x(), particle.pose.position.y - pose.y()));
+    particle.weight = weight_manager_.normal_pdf(distance, is_rtk_fixed);
+  }
+
+  return weighted_particles;
+}
+
+void GnssParticleCorrector::add_weight_by_orientation(
+  ParticleArray & weighted_particles, const Eigen::Vector3f & velocity)
+{
+  if (velocity.norm() < 1) return;
+
+  const float theta = std::atan2(velocity.x(), velocity.y());
+  const Eigen::Quaternionf measured(std::cos(theta / 2), 0, 0, std::sin(theta / 2));
+
+  auto pdf = [](float x) -> float {
+    constexpr float k = 3.0f;  // TODO:
```

Comment:
TODO?

Reply:
This function has been erased, so it's no longer a problem. :bow: